File: code-challenges/Data-Structures/stacksAndQueues/stacks-and-queues.js
class Node {
  constructor (value = null) {
    this.value = value
    this.next = null
  }
} 

class Stack {
  constructor () {
    this.top = null
  }

  peek() {
    return this.top.value
  }

  push(value) {
    let newNode = new Node(value)
    newNode.next = this.top
    this.top = newNode
    return this.top
  }

  pop() {
    if (!this.top) {
      return undefined
    } else if (!this.top.next) {
      this.top = null
      return this.top
    } else {
      this.top.value = this.top.next.value
      this.top.next = this.top.next.next
      return this.top.value
    }
  }

  isEmpty() {
    return !this.top ? true : false
  }
}

class Queue {
  constructor () {
    this.front = null
  }

  enqueue(value) {
    let newNode = new Node(value)
    if (!this.front) {
      this.front = newNode
    } else {
      let currNode = this.front
      while (currNode.next) {
        currNode = currNode.next
      }
      currNode.next = newNode
    }
    return this.head
  }

  dequeue() {
    let dequeueVal
    if (!this.front) {
      return undefined
    } else if (!this.front.next) {
      dequeueVal = this.front.value
      this.front = null
    } else {
      dequeueVal = this.front.value
      this.front.value = this.front.next.value
      this.front.next = this.front.next.next
    }
    return dequeueVal
  }

  peek() {
    return this.front.value
  }

  isEmpty() {
    return !this.front ? true : false
  }
}

module.exports = {
  Node,
  Stack,
  Queue
}

File: code_challenges/data_structures/stacks_and_queues/stacks_and_queues.py
class Node:
    def __init__(self, value, next = None):
        self.value = value
        self.next = next

class Stack:
    def __init__(self):
        self.top = None

    def peek(self):
        return self.top.value

    def push(self, val):
        new_node = Node(val)
        new_node.next = self.top
        self.top = new_node
        return self.top

    def pop(self):
        if not self.top:
            return None
        elif not self.top.next:
            popped = self.top
            self.top = None
            return popped
        else:
            popped = self.top.value
            self.top.value = self.top.next.value
            self.top.next = self.top.next.next
            return popped

    def isEmpty(self):
        return not self.top

    def __str__(self):
        stack = str()
        curr_node = self.top
        if not curr_node: return 'None'
        while curr_node.next:
            stack += str(curr_node.value) + ' -> '
            curr_node = curr_node.next
        stack += str(curr_node.value) + ' -> None'
        return stack

class Queue:
    def __init__(self):
        self.front = None

    def enqueue(self, val):
        new_node = Node(val)
        if not self.front:
            self.front = new_node
        else:
            curr_node = self.front
            while curr_node.next:
                curr_node = curr_node.next
            curr_node.next = new_node
        return self.front

    def dequeue(self):
        dequeue_value = None
        if not self.front:
            return dequeue_value
        elif not self.front.next:
            dequeue_value = self.front.value
            self.front = None
        else:
            dequeue_value = self.front.value
            self.front.value = self.front.next.value
            self.front.next = self.front.next.next
        return dequeue_value

    def peek(self):
        return self.front.value

    def isEmpty(self):
        return not self.front

    def __str__(self):
        queue = str()
        curr_node = self.front
        if not curr_node: return 'None'
        while curr_node.next:
            queue += str(curr_node.value) + ' -> '
            curr_node = curr_node.next
        queue += str(curr_node.value) + ' -> None'
        return queue

class PseudoQueue(Stack):
  def __init__(self):
      super().__init__()
      self.top = None

  def enqueue(self, value):
    first_stack = PseudoQueue()
    second_stack = PseudoQueue()
    if not self.top:
        second_stack.push(value)
    else:
      curr_node = self.top
      print(self)
      while curr_node.next:
        first_stack.push(curr_node.value)
        curr_node = curr_node.next
      first_stack.push(curr_node.value)
      first_stack.push(value)
      print(first_stack)
      other_node = first_stack.top
      while other_node.next:
        second_stack.push(other_node.value)
        other_node = other_node.next
    second_stack.push(other_node.value)
    self.top = second_stack.top
    return self.top

  def dequeue(self):
    return self.pop()

def __str__(self):
    pass

if __name__ == "__main__":
  test_pseudo = PseudoQueue()
  test_pseudo.push(1)
  test_pseudo.push(2)
  test_pseudo.push(3)
  test_pseudo.enqueue(4)
  print(test_pseudo)
